src/tree/updater_approx.h

```diff
@@ -0,0 +1,146 @@
+/*!
+ * Copyright 2021 XGBoost contributors
+ *
+ * \brief Implementation for the approx tree method.
+ */
+#ifndef XGBOOST_TREE_UPDATER_APPROX_H_
+#define XGBOOST_TREE_UPDATER_APPROX_H_
+
+#include <limits>
+#include <utility>
+#include <vector>
+
+#include "../common/partition_builder.h"
+#include "../common/random.h"
+#include "constraints.h"
+#include "driver.h"
+#include "hist/evaluate_splits.h"
+#include "hist/expand_entry.h"
+#include "hist/param.h"
+#include "param.h"
+#include "xgboost/json.h"
+#include "xgboost/tree_updater.h"
+
+namespace xgboost {
+namespace tree {
+class ApproxRowPartitioner {
+  static constexpr size_t kPartitionBlockSize = 2048;
+  common::PartitionBuilder<kPartitionBlockSize> partition_builder_;
+  common::RowSetCollection row_set_collection_;
+
+ public:
+  bst_row_t base_rowid = 0;
+
+  static auto SearchCutValue(bst_row_t ridx, bst_feature_t fidx, GHistIndexMatrix const &index,
+                             std::vector<uint32_t> const &cut_ptrs,
+                             std::vector<float> const &cut_values) {
+    int32_t gidx = -1;
+    auto const &row_ptr = index.row_ptr;
+    auto get_row_ptr = [&](size_t ridx) { return row_ptr.at(ridx - index.base_rowid); };
+
```

Comment:
This function looks the same as other places in the code base. Can (or should) it be extracted?

Reply:
Right now there's one similar function in histogram builder, but that one is specialized over first page as it's in the hot loop of building histogram.